rig/place_and_route/place/utils.py

```diff
@@ -86,3 +91,151 @@ def apply_reserve_resource_constraint(machine, constraint):
+
+    def __init__(self, vertices):
+        self.vertices = list(vertices)
+
+    def __repr__(self):
+        return "{}({})".format(self.__class__.__name__, repr(self.vertices))
+
+
+def apply_same_chip_constraints(vertices_resources, nets, constraints):
+    """Modify a set of vertices_resources, nets and constraints to account for
+    all SameChipConstraints.
+
+    To allow placement algorithms to handle SameChipConstraints without any
+    special cases, Vertices identified in a SameChipConstraint are merged into
+    a new vertex whose vertices_resources are the sum total of their parts
+    which may be placed as if a single vertex. Once placed, the placement can
+    be expanded into a full placement of all the original vertices using
+    :py:func:`finalise_same_chip_constraints`.
+
+    A typical use pattern might look like::
+
+        def my_placer(vertices_resources, nets, machine, constraints):
+            # Should be done first thing since this may redefine
+            # vertices_resources, nets and constraints.
+            vertices_resources, nets, constraints, substitutions = \\
+                apply_same_chip_constraints(vertices_resources,
+                                            nets, constraints)
+
+            # ...deal with other types of constraint...
+
+            # ...perform placement...
+
+            finalise_same_chip_constraints(substitutions, placements)
+            return placements
+
+    Note that this function does not modify its arguments but rather returns
+    new copies of the structures supplied.
+
+    Parameters
+    ----------
+    vertices_resources : {vertex: {resource: quantity, ...}, ...}
+    nets : [:py:class:`~rig.netlist.Net`, ...]
+    constraints : [constraint, ...]
+
+    Returns
+    -------
+    (vertices_resources, nets, constraints, substitutions)
+        The vertices_resources, nets and constraints values contain modified
+        copies of the supplied data structures modified to contain a single
+        vertex in place of the individual constrained vertices.
+
+        substitutions is a list of :py:class:`MergedVertex` objects which
+        resulted from the combining of the constrained vertices. The order of
+        the list is the order the substitutions were carried out. The
+        :py:func:`finalise_same_chip_constraints` function can be used to
+        expand a set of substitutions.
+    """
+    # Make a copy of the basic structures to be modified by this function
+    vertices_resources = vertices_resources.copy()
+    nets = nets[:]
+    constraints = constraints[:]
+
+    substitutions = []
+
+    for same_chip_constraint in constraints:
+        if not isinstance(same_chip_constraint, SameChipConstraint):
+            continue
+
+        # Skip constraints which don't actually merge anything...
+        if len(same_chip_constraint.vertices) <= 1:
+            continue
+
+        # The new (merged) vertex to replace the constrained vertices with
```

Comment:
"The new (merged) vertex with which to replace the constrained vertices"?

"This is the type of arrant pedantry up with which I will not put"
:stuck_out_tongue: